File: sl/common/websocket.py
from aiohttp import web
from common import Logger
from common import Socket
from common import utils

class Websocket(Socket):
    def __init__(self, source=None, websocket=None):
        super(Websocket, self).__init__(source=source)
        self.__websocket = websocket

    async def listen(self):
        try:
            async for message in self.__websocket:
                if isinstance(self.__websocket, web.WebSocketResponse):
                    if message.type == WSMsgType.TEXT:
                        if message.data == 'close':
                            message = None
                        else:
                            message = message.data
                    elif msg.type == WSMsgType.ERROR:
                        message = None

                if message is not None:
                    message = utils.parse_data(message)
                    if message.get('type', None) is not None:
                        Logger.log_info('message')
                        Logger.log_info(message)

                    if (message.get('type', None) == 'service'
                            and message.get('action', None) is not None):
                        Logger.log_info('action accepted, trying to pass')
                        result = await self.execute(
                                message.get('action', None),
                                *message.get('args', []),
                                **message.get('kwargs', {}),)
                        response = {
                                'type':'service',
                                'id':message['id'],
                                'action_result':utils.pandas_to_dict(result),}
                        Logger.log_info('action resolved, trying to response')
                        await self.on_data(response)
                    else:
                        await self.push(message)
        except Exception as e:
            Logger.log_error(e)

        finally:
            await self.close()

    async def on_data(self, data):
        if data.get('type', None) is not None:
            Logger.log_info('message out')
            Logger.log_info(data)
        data = utils.stringify_data(data)
        if isinstance(self.__websocket, web.WebSocketResponse):
            await self.__websocket.send_str(data)
        else:
            await self.__websocket.send(data)

File: sl/sv/static/js/websocket.js
import Socket from './socket.js'

export default class extends Socket {
    constructor({source=undefined, websocket=undefined}){
        super({source:source});
        this.__websocket = websocket;
        this.__websocket.onmessage = this.listen;}

    async listen(event){
        var message = event.data;
        var action_result;

        if (message){
            message = JSON.parse(message);
            if (message['type'] == 'service' && message['action']){
                action_result = await self.execute(
                        message['action'],
                        message['args'],
                        message['kwargs'],);
                await this.on_data(action_result);}
            else {
                await this.push(message);}}}
    
    async on_data(data){
        data = JSON.stringify(data);
        await this.__websocket.send(data);}}
